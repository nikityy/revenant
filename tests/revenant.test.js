const Revenant = require("../lib/revenant");
const {
  InvalidCredentialsError,
  NotAuthorizedError
} = require("../lib/errors");
const RutrackerMock = require("./mocks/rutracker-mock");
const ConfigMock = require("./mocks/config-mock");
const { getHashedSnapshot } = require("./utils");

const { VALID_CREDENTIALS, INVALID_CREDENTIALS, RESULTS } = RutrackerMock;
const { WATCH_LIST } = ConfigMock;

const REVENANT_CONFIG = {
  configPath: ""
};

describe("#login", () => {
  test("adds valid credentials to config", () => {
    expect.assertions(2);

    const setCookieMock = jest.fn().mockResolvedValue(true);
    const config = new ConfigMock();
    config.setCookie = setCookieMock;

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return revenant.login(VALID_CREDENTIALS).then(() => {
      expect(setCookieMock).toHaveBeenCalledTimes(1);
      expect(setCookieMock).toHaveBeenCalledWith(RutrackerMock.COOKIE);
    });
  });

  test("rejects if credentials are invalid", () => {
    expect.assertions(1);

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.rutracker = new RutrackerMock();

    return expect(revenant.login(INVALID_CREDENTIALS)).rejects.toThrow(
      InvalidCredentialsError
    );
  });
});

describe("#addToWatchList", () => {
  test("adds item to config", () => {
    expect.assertions(2);

    const setWatchListMock = jest.fn().mockResolvedValue(true);
    const config = new ConfigMock();
    config.setWatchList = setWatchListMock;

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return revenant.addToWatchList("D").then(() => {
      expect(setWatchListMock).toHaveBeenCalledTimes(1);
      expect(setWatchListMock).toHaveBeenCalledWith([...WATCH_LIST, "D"]);
    });
  });

  test("does nothing if item is already in list", () => {
    expect.assertions(1);

    const setWatchListMock = jest.fn().mockResolvedValue(true);
    const config = new ConfigMock();
    config.setWatchList = setWatchListMock;

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return revenant.addToWatchList(WATCH_LIST[0]).then(() => {
      expect(setWatchListMock).toHaveBeenCalledTimes(0);
    });
  });
});

describe("#removeFromWatchList", () => {
  test("removes item from config", () => {
    expect.assertions(2);

    const setWatchListMock = jest.fn().mockResolvedValue(true);
    const config = new ConfigMock();
    config.setWatchList = setWatchListMock;

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return revenant.removeFromWatchList("B").then(() => {
      expect(setWatchListMock).toHaveBeenCalledTimes(1);
      expect(setWatchListMock).toHaveBeenCalledWith(
        WATCH_LIST.filter(x => x !== "B")
      );
    });
  });

  test("does nothing if item is not in list", () => {
    expect.assertions(1);

    const setWatchListMock = jest.fn().mockResolvedValue(true);
    const config = new ConfigMock();
    config.setWatchList = setWatchListMock;

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return revenant.removeFromWatchList("D").then(() => {
      expect(setWatchListMock).toHaveBeenCalledTimes(0);
    });
  });
});

describe("#getUpdates", () => {
  test("resolves with updated items", () => {
    expect.assertions(1);

    const config = new ConfigMock();

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return expect(revenant.getUpdates()).resolves.toEqual({
      A: RESULTS.A,
      B: RESULTS.B,
      C: RESULTS.C
    });
  });

  test("adds updated items to config", () => {
    expect.assertions(2);

    const setSnapshotsMock = jest.fn().mockResolvedValue(true);
    const config = new ConfigMock();
    config.setSnapshots = setSnapshotsMock;

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return revenant.getUpdates().then(() => {
      expect(setSnapshotsMock).toHaveBeenCalledTimes(1);
      expect(setSnapshotsMock).toHaveBeenCalledWith({
        A: getHashedSnapshot(RESULTS.A),
        B: getHashedSnapshot(RESULTS.B),
        C: getHashedSnapshot(RESULTS.C)
      });
    });
  });

  test("rejects if not authorized", () => {
    expect.assertions(1);

    const config = new ConfigMock();
    config.getCookie = jest.fn().mockResolvedValue(null);

    const revenant = new Revenant(REVENANT_CONFIG);
    revenant.config = config;
    revenant.rutracker = new RutrackerMock();

    return expect(revenant.getUpdates()).rejects.toThrow(NotAuthorizedError);
  });
});
