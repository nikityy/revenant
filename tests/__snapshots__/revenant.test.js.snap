// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list should print all watchlist entries 1`] = `
Array [
  "Маленькие женщины Грета Гервиг",
  "Дау Илья Хржановский",
  "Довод Кристофер Нолан",
  "1917 Сэм Мендес",
  "Джентльмены Гай Ричи",
  "Офицер и шпион Роман Полански",
  "Проверка на дорогах Алексей Герман",
]
`;

exports[`list should print all watchlist entries 2`] = `
Object {
  "downloadPath": "~/Downloads/",
  "kinopoisk": Object {
    "watchUrl": "https://www.kinopoisk.ru/user/789114/movies/list/type/3575/#list",
  },
  "rutracker": Object {
    "cookie": "bb-cookie:123",
  },
  "torrents": Object {
    "1917 Сэм Мендес": Array [],
    "Дау Илья Хржановский": Array [],
    "Джентльмены Гай Ричи": Array [],
    "Довод Кристофер Нолан": Array [],
    "Маленькие женщины Грета Гервиг": Array [],
    "Офицер и шпион Роман Полански": Array [],
    "Проверка на дорогах Алексей Герман": Array [],
  },
  "version": "1",
  "watchlist": Array [
    "Маленькие женщины Грета Гервиг",
    "Дау Илья Хржановский",
    "Довод Кристофер Нолан",
    "1917 Сэм Мендес",
    "Джентльмены Гай Ричи",
    "Офицер и шпион Роман Полански",
    "Проверка на дорогах Алексей Герман",
  ],
}
`;
